File: SEM5/FLCD/lab2/lab_ST_all/src/HashTable.java
import java.util.ArrayList;

public class HashTable implements SymbolTableInterface {
    private Integer size;
    private ArrayList<ArrayList<String>> table;

    public HashTable(Integer size){
        this.size = size;
        this.table = new ArrayList<>();
        for(int i = 0; i < size; i++){
            this.table.add(new ArrayList<>());
        }
    }

    @Override
    public String findByPos(Pair pos){
        if(this.table.size() <= pos.getFirst() || this.table.get(pos.getFirst()).size() <= pos.getSecond()){
            throw new IndexOutOfBoundsException("Invalid position");
        }
        return this.table.get(pos.getFirst()).get(pos.getSecond());
    }

    @Override
    public Integer getSize(){
        return size;
    }

    @Override
    public Pair findPositionOfTerm(String term){
        int pos = hash(term);
        if(!table.get(pos).isEmpty()){
            ArrayList<String> elems = this.table.get(pos);
            for(int i = 0; i < elems.size(); i++){
                if(elems.get(i).equals(term)){
                    return new Pair(pos, i);
                }
            }
        }
        return null;
    }

    private Integer hash(String key){
        int sumChars = 0;
        for(char c : key.toCharArray()){
            sumChars += c;
        }
        return sumChars % size;
    }

    @Override
    public boolean containsTerm(String term){
        return this.findPositionOfTerm(term) != null;
    }

    @Override
    public boolean add(String term){
        if(containsTerm(term)){
            return false;
        }
        Integer pos = hash(term);
        ArrayList<String> elems = this.table.get(pos);
        elems.add(term);
        return true;
    }

    @Override
    public String toString() {
        return "HashTable { elements=" + table + ", size=" + size + " }";
    }
}

File: SEM5/FLCD/lab2/lab2_ST_cpp/lab2_ST_cpp/HashTable.cpp
#include "HashTable.h"
#include <stdexcept>
#include <sstream>

// Hash function
int HashTable::hash(const std::string& key) {
    int sumChars = 0;
    for (char c : key) {
        sumChars += c;
    }
    return sumChars % size;
}

// Constructor
HashTable::HashTable(int size) {
    this->size = size;  // Fix: Use this->size to assign to the member variable
    hashTable.resize(size);
}

// Find by position
std::string HashTable::findByPosition(Pair position) {
    if (hashTable.size() <= position.getFirst() || hashTable[position.getFirst()].size() <= position.getSecond()) {
        throw std::out_of_range("Invalid position");
    }
    return hashTable[position.getFirst()][position.getSecond()];
}

// Get size
int HashTable::getSize() {
    return size;
}

// Find position of a term
Pair HashTable::findPositionOfTerm(const std::string& term) {
    int position = hash(term);
    if (!hashTable[position].empty()) {
        for (int i = 0; i < hashTable[position].size(); i++) {
            if (hashTable[position][i] == term)
                return Pair(position, i);
        }
    }
    return Pair(-1, -1);
}

// Check if a term exists
bool HashTable::containsTerm(const std::string& term) {
    Pair pos = findPositionOfTerm(term);
    return pos.getFirst() != -1 && pos.getSecond() != -1;
}

// Add a term
bool HashTable::add(const std::string& term) {
    if (containsTerm(term)) {
        return false;
    }
    int pos = hash(term);
    hashTable[pos].push_back(term);
    return true;
}

// ToString equivalent
std::string HashTable::toString() {
    std::stringstream ss;
    ss << "SymbolTable { elements= [";
    for (int i = 0; i < hashTable.size(); ++i) {
        ss << "[";
        for (int j = 0; j < hashTable[i].size(); ++j) {
            ss << hashTable[i][j];
            if (j != hashTable[i].size() - 1) {
                ss << ", ";
            }
        }
        ss << "]";
        if (i != hashTable.size() - 1) {
            ss << ", ";
        }
    }
    ss << "], size = " << size << " }";
    return ss.str();
}
